File: mobile/lib/services/ikuai_client.dart
/// 爱快（iKuai）路由器 API 客户端（Dart 移植版）。
///
/// 由 Web 版 `web/app/ikuai_client.py` 完整移植，行为保持一致：
/// 登录认证（MD5 + Base64 + salt 多版本重试）、会话管理、通用调用、
/// DHCP 静态分配列表读取、网关 A/B 切换、字段兼容。
///
/// 爱快 API 关键结论（联调实测）：
/// - 登录：`POST /Action/login`，body `{username, passwd: MD5(密码),
///   pass: Base64(salt+密码), remember_password: null}`。
///   salt 前缀随固件变化（salt_123 / salt_113 / salt_11 / salt_13 …），逐个尝试。
/// - 业务：`POST /Action/call`，body `{func_name, action, param}`。
///   **成功判断（关键坑）**：不是只看 `Result` 码，而是
///   `Result == 10000`（旧版）或 `code == 0`（企业版 4.x）或 `ErrMsg == "Success"`。
/// - DHCP 静态分配：`func_name=dhcp_static`，读用
///   `TYPE=static_total,static_data`（配置视角，含网关字段）；
///   写用 edit/add，param 为平铺字段，缺 `enabled` 会报「参数错误」。
/// - 会话：登录后 `sess_key` cookie；会话过期码 10001，自动重登重试一次。
library;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;
import 'package:http/io_client.dart';

import '../models/models.dart';

/// 爱快 API 异常，携带错误码与可读信息。
class IkuaiError implements Exception {
  IkuaiError(this.code, this.message);

  /// 错误码：401 认证失败 / 404 记录不存在 / 503 网络不可达 / 其他为爱快返回码。
  final int code;

  /// 面向用户的可读信息。
  final String message;

  @override
  String toString() => '[$code] $message';
}

// ---- 认证编码工具（纯函数，便于单元测试）----

/// `MD5(明文密码)`，返回 32 位小写十六进制。
String md5Password(String password) =>
    md5.convert(utf8.encode(password)).toString();

/// `Base64(salt 前缀 + 明文密码)`。
String encodePass(String password, String salt) =>
    base64.encode(utf8.encode('$salt$password'));

/// 返回 `[passwd, pass]`，供登录请求体使用。
List<String> encodeCredentials(String password, String salt) =>
    [md5Password(password), encodePass(password, salt)];

// ---- 网关 A/B 决策（纯函数）----

/// 当前网关 == B → 切到 A；其余（含无法匹配）按「当前≈A」处理，切到 B。
String decideTargetGateway(String current, String gatewayA, String gatewayB) =>
    current == gatewayB ? gatewayA : gatewayB;

// ---- 字段兼容工具 ----

/// 按优先级取字段，返回非空字符串；兼容爱快多字段命名。
String getField(Map<String, dynamic> item, List<String> keys) {
  for (final key in keys) {
    final value = item[key];
    if (value == null) continue;
    final text = value.toString().trim();
    if (text.isNotEmpty) return text;
  }
  return '';
}

/// 爱快 API 客户端（网络层），只负责「与爱快通信」，不掺杂业务判断。
class IkuaiClient {
  IkuaiClient({
    required this.host,
    this.port = 80,
    this.username = '',
    this.password = '',
    this.useHttps = false,
    List<String>? saltList,
    this.timeout = const Duration(seconds: 20),
    http.Client? transport,
  }) : saltList = saltList ?? defaultSaltList,
       _transport = transport ?? buildDefaultTransport(timeout);

  static const List<String> defaultSaltList = [
    'salt_123',
    'salt_113',
    'salt_11',
    'salt_13',
  ];

  static const int successCode = 10000;
  static const int sessionExpiredCode = 10001;

  /// 默认传输层：忽略自签名 HTTPS 证书（爱快默认自签名）。
  static http.Client buildDefaultTransport(Duration timeout) {
    final inner = HttpClient();
    inner.badCertificateCallback = (cert, host, port) => true;
    inner.connectionTimeout = timeout;
    return IOClient(inner);
  }

  final String host;
  final int port;
  final String username;
  final String password;
  final bool useHttps;
  final List<String> saltList;
  final Duration timeout;

  final http.Client _transport;
  final Map<String, String> _cookies = {};

  bool _loggingIn = false;
  bool _loggedIn = false;

  /// 是否已完成登录（会话过期由 [call] 自动重登）。
  bool get isLoggedIn => _loggedIn;

  /// 基础地址，如 `http://192.168.1.1:80`。
  String get baseUrl => '${useHttps ? 'https' : 'http'}://$host:$port';

  /// 按当前配置构造客户端；便于上层在配置变更时重建。
  factory IkuaiClient.fromConfig(
    IkuaiConfig cfg, {
    http.Client? transport,
    List<String>? saltList,
  }) {
    return IkuaiClient(
      host: cfg.host.trim(),
      port: cfg.port,
      username: cfg.username,
      password: cfg.password,
      useHttps: cfg.useHttps,
      transport: transport,
      saltList: saltList,
    );
  }

  // ---- 登录 ----

  /// 仅在未登录时执行登录；已登录则跳过。
  Future<void> ensureLogin() async {
    if (!_loggedIn) await login();
  }

  /// 逐个 salt 尝试登录，全部失败则抛 [IkuaiError]（code = 401）。
  Future<void> login() async {
    if (_loggingIn) return;
    _loggingIn = true;
    try {
      final errors = <String>[];
      for (final salt in saltList) {
        try {
          await _tryLogin(salt);
          _loggedIn = true;
          return;
        } on IkuaiError catch (exc) {
          errors.add('[$salt] ${exc.message}');
        }
      }
      throw IkuaiError(401, '登录失败：${errors.join('；')}');
    } finally {
      _loggingIn = false;
    }
  }

  Future<void> _tryLogin(String salt) async {
    final credentials = encodeCredentials(password, salt);
    final body = <String, dynamic>{
      'username': username,
      'passwd': credentials[0],
      'pass': credentials[1],
      'remember_password': null,
    };
    final data = await _post('/Action/login', body);
    if (!isSuccess(data)) {
      throw IkuaiError(
        extractResult(data),
        extractErrMsg(data) ?? '登录失败',
      );
    }
    _storeSession(data);
  }

  // ---- 通用调用 ----

  /// 调用爱快业务接口；识别到会话过期会自动重登并重试一次。
  Future<dynamic> call(
    String funcName,
    String action, [
    Map<String, dynamic>? param,
  ]) async {
    final body = <String, dynamic>{
      'func_name': funcName,
      'action': action,
      'param': param ?? <String, dynamic>{},
    };
    var data = await _post('/Action/call', body);
    if (isSessionExpired(data)) {
      _loggedIn = false;
      _cookies.clear();
      await login();
      data = await _post('/Action/call', body);
    }
    if (!isSuccess(data)) {
      throw IkuaiError(
        extractResult(data),
        extractErrMsg(data) ?? '调用失败（原始返回：${summarize(data)}）',
      );
    }
    return data;
  }

  Future<dynamic> _post(String path, Map<String, dynamic> body) async {
    final uri = Uri.parse('$baseUrl$path');
    final headers = <String, String>{'Content-Type': 'application/json'};
    if (_cookies.isNotEmpty) {
      headers['Cookie'] = _cookies.entries
          .map((e) => '${e.key}=${e.value}')
          .join('; ');
    }

    http.Response response;
    try {
      response = await _transport
          .post(uri, headers: headers, body: jsonEncode(body))
          .timeout(timeout);
    } on TimeoutException {
      throw IkuaiError(503, '无法连接路由器：请求超时（${timeout.inSeconds} 秒）');
    } on IkuaiError {
      rethrow;
    } catch (exc) {
      throw IkuaiError(503, '无法连接路由器：$exc');
    }

    _captureCookies(response);
    return toJson(response);
  }

  // ---- DHCP 静态分配 ----

  /// 读取 DHCP 静态分配列表（原始字段，含 gateway/enabled/dns）。
  ///
  /// 注意：必须用 `TYPE=static_total,static_data`（前端同款参数），
  /// 才能拿到配置视角字段（gateway/enabled/dns1/dns2）；
  /// `TYPE=total,data` 是 ARP/租约视角，不含网关字段。
  Future<List<Map<String, dynamic>>> getDhcpBindings() async {
    final data = await call('dhcp_static', 'show', {
      'TYPE': 'static_total,static_data',
      'limit': '0,500',
    });
    return extractList(data);
  }

  /// 新增或编辑一条 DHCP 静态绑定记录（平铺字段）。
  Future<void> saveDhcpBinding(
    Map<String, dynamic> item, {
    required bool isEdit,
  }) async {
    await call('dhcp_static', isEdit ? 'edit' : 'add', item);
  }

  // ---- 切换网关（业务编排）----

  /// 切换指定终端的网关，返回原网关与新网关。
  Future<ToggleResult> toggleGateway(
    String mac,
    String gatewayA,
    String gatewayB,
  ) async {
    final bindings = await getDhcpBindings();
    Map<String, dynamic>? target;
    for (final item in bindings) {
      if (getField(item, ['mac']).toLowerCase() == mac.toLowerCase()) {
        target = item;
        break;
      }
    }
    if (target == null) {
      throw IkuaiError(404, '未找到该终端的静态绑定记录（MAC: $mac）');
    }

    final current = getField(target, ['gateway', 'gw']);
    final newGateway = decideTargetGateway(current, gatewayA, gatewayB);

    // edit 用配置视角的干净字段集（前端同款）；enabled 缺失会导致「参数错误」
    final rawEnabled = target['enabled'];
    final enabled =
        (rawEnabled == null ||
            rawEnabled == false ||
            rawEnabled.toString().trim().isEmpty)
        ? 'yes'
        : rawEnabled;

    final updated = <String, dynamic>{
      'id': target['id'],
      'enabled': enabled,
      'interface': target['interface'] ?? '',
      'mac': getField(target, ['mac']),
      'ip_addr': getField(target, ['ip_addr']),
      'gateway': newGateway,
      'dns1': target['dns1'] ?? '',
      'dns2': target['dns2'] ?? '',
      'comment': target['comment'] ?? '',
    };
    await saveDhcpBinding(updated, isEdit: true);
    return ToggleResult(
      mac: mac,
      oldGateway: current,
      newGateway: newGateway,
    );
  }

  /// 释放底层连接（页面销毁时调用）。
  void dispose() => _transport.close();

  // ---- 会话与解析工具 ----

  void _captureCookies(http.Response response) {
    final raw = response.headers['set-cookie'];
    if (raw == null || raw.isEmpty) return;
    for (final piece in raw.split(',')) {
      final segment = piece.split(';').first.trim();
      final index = segment.indexOf('=');
      if (index <= 0) continue;
      _cookies[segment.substring(0, index)] = segment.substring(index + 1);
    }
  }

  /// 优先用 Set-Cookie（已自动捕获）；若响应体含 sess_key 则补存。
  void _storeSession(dynamic data) {
    if (data is Map) {
      for (final key in ['sess_key', 'SessionKey', 'session_key', 'sesskey']) {
        final value = data[key];
        if (value != null && value.toString().isNotEmpty) {
          _cookies['sess_key'] = value.toString();
          return;
        }
      }
    }
  }

  /// 解析响应体为 JSON；兼容带 `sending to kernel` 前缀的固件响应。
  static dynamic toJson(http.Response response) {
    final text = response.body;
    try {
      return jsonDecode(text);
    } on FormatException {
      var cleaned = text;
      if (cleaned.startsWith('sending to kernel')) {
        cleaned = cleaned.replaceFirst('sending to kernel', '').trim();
      }
      try {
        return jsonDecode(cleaned);
      } on FormatException {
        return {'raw': text};
      }
    }
  }

  /// 提取返回码（Result / result / code / Code / ret），无法解析返回 -1。
  static int extractResult(dynamic data) {
    if (data is Map) {
      for (final key in ['Result', 'result', 'code', 'Code', 'ret']) {
        if (!data.containsKey(key)) continue;
        final value = data[key];
        if (value is bool) continue;
        if (value is int) return value;
        if (value is num) return value.toInt();
        if (value is String) {
          final parsed = int.tryParse(value);
          if (parsed != null) return parsed;
        }
      }
    }
    return -1;
  }

  /// 成功判断：`Result == 10000` / `code == 0` / `ErrMsg == "Success"` 三者任一。
  static bool isSuccess(dynamic data) {
    final result = extractResult(data);
    if (result == successCode || result == 0) return true;
    final msg = extractErrMsg(data);
    return msg != null && msg.trim().toLowerCase() == 'success';
  }

  /// 提取错误信息字段。
  static String? extractErrMsg(dynamic data) {
    if (data is Map) {
      for (final key in [
        'ErrMsg',
        'errmsg',
        'ErrorMsg',
        'error_msg',
        'message',
        'msg',
      ]) {
        if (data.containsKey(key) && data[key] != null) {
          return data[key].toString();
        }
      }
    }
    return null;
  }

  /// 是否为「无法解析为 JSON」的响应。
  ///
  /// [toJson] 遇到非 JSON 文本时会包装成 `{'raw': <原始文本>}`，
  /// 技术方案约定这类响应视为会话失效，需重新登录。
  static bool isUnparsedResponse(dynamic data) =>
      data is Map && data.length == 1 && data.containsKey('raw');

  /// 会话是否已过期（返回码 10001、响应非 JSON 或提示语含未登录/过期字样）。
  static bool isSessionExpired(dynamic data) {
    if (data is! Map) return true;
    if (isUnparsedResponse(data)) return true;
    if (extractResult(data) == sessionExpiredCode) return true;
    final msg = (extractErrMsg(data) ?? '').toLowerCase();
    return const ['no login', '未登录', 'expired', '过期']
        .any((k) => msg.contains(k));
  }

  /// 截断后的原始返回，便于排查。
  static String summarize(dynamic data) {
    String text;
    try {
      text = jsonEncode(data);
    } catch (_) {
      text = data.toString();
    }
    return text.length <= 200 ? text : '${text.substring(0, 200)}…';
  }

  /// 从响应中提取列表数据，兼容旧版 `Data` 与企业版 `results` 多层容器。
  static List<Map<String, dynamic>> extractList(dynamic data) {
    dynamic found;
    if (data is List) {
      found = data;
    } else if (data is Map) {
      dynamic container;
      for (final key in [
        'Data',
        'data',
        'result',
        'results',
        'static_data',
        'list',
        'items',
        'rows',
      ]) {
        if (data.containsKey(key)) {
          container = data[key];
          break;
        }
      }
      if (container is List) {
        found = container;
      } else if (container is Map) {
        for (final key in [
          'data',
          'result',
          'results',
          'static_data',
          'list',
          'items',
          'rows',
        ]) {
          if (container[key] is List) {
            found = container[key];
            break;
          }
        }
      }
    }
    if (found is! List) return const [];
    return found
        .whereType<Map>()
        .map((e) => Map<String, dynamic>.from(e))
        .toList();
  }
}

File: web/app/ikuai_client.py
"""爱快（iKuai）路由器 API 客户端（Python 移植版）。

将原 Flutter 版 `IkuaiClient` 的核心逻辑移植为 Python 实现：
登录认证（MD5 + Base64 + salt 多版本重试）、会话管理、通用调用、
DHCP 静态分配列表读取、网关 A/B 切换、字段兼容。

爱快 API 关键结论（联调实测）：
- 登录：`POST /Action/login`，body `{username, passwd: MD5(密码), pass: Base64(salt+密码), remember_password: null}`。
  - 字段是 `username`（非 user_name）；`passwd`=md5(密码)，`pass`=base64(salt前缀+密码)。
  - salt 前缀随固件变化（salt_123 / salt_113 / salt_11 / salt_13 …），逐个尝试。
- 业务：`POST /Action/call`，body `{func_name, action, param}`。
  - **成功判断（关键坑）**：不是看 `Result` 码，而是 `ErrMsg=="Success"`（旧版）或 `code==0`（企业版 4.x）；
    部分固件 `/Action/call` 响应不含 `Result` 字段，需综合判断 `Result==10000 || code==0 || ErrMsg=="Success"`。
  - 列表数据容器：旧版 `Data`、企业版 `results`；内层字段 `data`/`result` 等，需多字段兼容。
- DHCP 静态绑定：`func_name=dhcp_static`（不是 `dhcp_addr_bind`）；读=show(`{"TYPE":"total,data","limit":"0,500"}`)；
  写=add/edit（param 为平铺字段，不要 `{"data":{}}` 包装）。字段：mac、ip_addr、comment、dns1/dns2（3.7.12+）。
  - **网关字段名待确认**：爱快不同模块命名不一（`static_rt` 用 `gateway`，DHCP 相关常缩写 `gw`），
    逆向项目 gxxHuang 的 dhcp_static `add` 未覆盖网关字段，本实现默认用 `gw` 并做双字段兼容。
- 会话：登录后 `sess_key` cookie；会话过期码 10001，自动重登重试一次。
"""
from __future__ import annotations

import base64
import hashlib
import json
from typing import Any, Optional

import requests
import urllib3

# 忽略自签名证书告警（爱快默认自签名 HTTPS）
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)


class IkuaiError(Exception):
    """爱快 API 异常，携带错误码与可读信息。"""

    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(f"[{code}] {message}")


# ---- 认证编码工具（纯函数，便于单元测试）----


def md5_password(password: str) -> str:
    """`MD5(明文密码)`，返回 32 位小写十六进制。"""
    return hashlib.md5(password.encode("utf-8")).hexdigest()


def encode_pass(password: str, salt: str) -> str:
    """`Base64(salt 前缀 + 明文密码)`。"""
    return base64.b64encode(f"{salt}{password}".encode("utf-8")).decode("ascii")


def encode_credentials(password: str, salt: str) -> tuple[str, str]:
    """返回 `(passwd, pass)` 元组，供登录请求体使用。"""
    return md5_password(password), encode_pass(password, salt)


# ---- 网关 A/B 决策（纯函数）----


def decide_target_gateway(current: str, gateway_a: str, gateway_b: str) -> str:
    """当前网关 == B → 切到 A；其余（含无法匹配）按「当前≈A」处理，切到 B。"""
    if current == gateway_b:
        return gateway_a
    return gateway_b


# ---- 字段兼容工具 ----


def get_field(item: dict, *keys: str) -> str:
    """按优先级取字段，返回非空字符串；兼容爱快多字段命名。"""
    for key in keys:
        value = item.get(key)
        if value is not None and str(value).strip() != "":
            return str(value).strip()
    return ""


class IkuaiClient:
    """爱快 API 客户端（网络层），只负责「与爱快通信」，不掺杂业务判断。"""

    DEFAULT_SALT_LIST = ["salt_123", "salt_113", "salt_11", "salt_13"]
    SUCCESS_CODE = 10000
    SESSION_EXPIRED_CODE = 10001

    def __init__(
        self,
        host: str,
        port: int = 80,
        username: str = "",
        password: str = "",
        use_https: bool = False,
        salt_list: Optional[list[str]] = None,
        timeout: int = 20,
    ):
        self.host = host
        self.port = int(port)
        self.username = username
        self.password = password
        self.use_https = use_https
        self.salt_list = salt_list or self.DEFAULT_SALT_LIST
        self.timeout = timeout

        self.session = requests.Session()
        self.session.verify = False  # 忽略自签名证书
        self._logging_in = False
        self._logged_in = False

    # ---- 基础属性 ----

    @property
    def base_url(self) -> str:
        scheme = "https" if self.use_https else "http"
        return f"{scheme}://{self.host}:{self.port}"

    # ---- 登录 ----

    def ensure_login(self) -> None:
        """仅在未登录时执行登录；已登录则跳过（会话过期由 call 自动重登）。"""
        if not self._logged_in:
            self.login()

    def login(self) -> None:
        if self._logging_in:
            return
        self._logging_in = True
        try:
            errors: list[str] = []
            for salt in self.salt_list:
                try:
                    self._try_login(salt)
                    self._logged_in = True
                    return
                except IkuaiError as exc:
                    errors.append(f"[{salt}] {exc.message}")
            raise IkuaiError(401, "登录失败：" + "；".join(errors))
        finally:
            self._logging_in = False

    def _try_login(self, salt: str) -> None:
        passwd, pass_ = encode_credentials(self.password, salt)
        body = {
            "username": self.username,
            "passwd": passwd,
            "pass": pass_,
            "remember_password": None,
        }
        data = self._post("/Action/login", body)
        if not self._is_success(data):
            raise IkuaiError(
                self._extract_result(data),
                self._extract_err_msg(data) or "登录失败",
            )
        self._store_session(data)

    # ---- 通用调用 ----

    def call(self, func_name: str, action: str, param: Optional[dict] = None) -> Any:
        body = {"func_name": func_name, "action": action, "param": param or {}}
        data = self._post("/Action/call", body)
        if self._is_session_expired(data):
            self._logged_in = False
            self.login()
            data = self._post("/Action/call", body)
        if not self._is_success(data):
            raise IkuaiError(
                self._extract_result(data),
                self._extract_err_msg(data) or f"调用失败（原始返回：{self._summarize(data)}）",
            )
        return data

    def _post(self, path: str, body: dict) -> Any:
        try:
            resp = self.session.post(f"{self.base_url}{path}", json=body, timeout=self.timeout)
        except requests.RequestException as exc:
            raise IkuaiError(503, f"无法连接路由器：{exc}") from exc
        return self._to_json(resp)

    # ---- DHCP 静态分配 ----

    def get_dhcp_bindings(self) -> list[dict]:
        """读取 DHCP 静态分配列表（原始字段）。"""
        data = self.call("dhcp_static", "show", {"TYPE": "total,data", "limit": "0,500"})
        return self._extract_list(data)

    def save_dhcp_binding(self, item: dict, is_edit: bool) -> None:
        """新增或编辑一条 DHCP 静态绑定记录（平铺字段）。"""
        self.call("dhcp_static", "edit" if is_edit else "add", item)

    # ---- 切换网关（业务编排）----

    def toggle_gateway(self, mac: str, gateway_a: str, gateway_b: str) -> tuple[str, str]:
        """切换指定终端的网关，返回 `(原网关, 新网关)`。"""
        bindings = self.get_dhcp_bindings()
        target: Optional[dict] = None
        for item in bindings:
            if get_field(item, "mac").lower() == mac.lower():
                target = item
                break
        if target is None:
            raise IkuaiError(404, f"未找到该终端的静态绑定记录（MAC: {mac}）")

        current = get_field(target, "gw", "gateway")
        new_gateway = decide_target_gateway(current, gateway_a, gateway_b)

        updated = dict(target)
        self._set_gateway_field(updated, new_gateway)
        self.save_dhcp_binding(updated, is_edit=True)
        return current, new_gateway

    @staticmethod
    def _set_gateway_field(item: dict, gateway: str) -> None:
        """设置网关字段。爱快 dhcp_static 网关字段名待确认（gw vs gateway），默认 `gw`。"""
        item.pop("gateway", None)
        item["gw"] = gateway

    # ---- 解析工具 ----

    @staticmethod
    def _to_json(resp: requests.Response) -> Any:
        try:
            return resp.json()
        except ValueError:
            text = resp.text or ""
            # 部分固件响应带 "sending to kernel" 前缀
            if text.startswith("sending to kernel"):
                text = text.replace("sending to kernel", "", 1).strip()
            try:
                return json.loads(text)
            except ValueError:
                return {"raw": text}

    @staticmethod
    def _extract_result(data: Any) -> int:
        if isinstance(data, dict):
            for key in ("Result", "result", "code", "Code", "ret"):
                if key not in data:
                    continue
                value = data[key]
                if isinstance(value, bool):
                    continue
                if isinstance(value, int):
                    return value
                if isinstance(value, float):
                    return int(value)
                if isinstance(value, str):
                    try:
                        return int(value)
                    except ValueError:
                        pass
        return -1

    @classmethod
    def _is_success(cls, data: Any) -> bool:
        result = cls._extract_result(data)
        if result in (cls.SUCCESS_CODE, 0):
            return True
        msg = cls._extract_err_msg(data)
        return msg is not None and msg.strip().lower() == "success"

    @staticmethod
    def _extract_err_msg(data: Any) -> Optional[str]:
        if isinstance(data, dict):
            for key in ("ErrMsg", "errmsg", "ErrorMsg", "error_msg", "message", "msg"):
                if key in data and data[key] is not None:
                    return str(data[key])
        return None

    @staticmethod
    def _summarize(data: Any) -> str:
        try:
            text = json.dumps(data, ensure_ascii=False)
        except (TypeError, ValueError):
            text = str(data)
        return text if len(text) <= 200 else text[:200] + "…"

    @classmethod
    def _is_session_expired(cls, data: Any) -> bool:
        if not isinstance(data, dict):
            return True
        if cls._extract_result(data) == cls.SESSION_EXPIRED_CODE:
            return True
        msg = (cls._extract_err_msg(data) or "").lower()
        return any(k in msg for k in ("no login", "未登录", "expired", "过期"))

    def _store_session(self, data: Any) -> None:
        """优先用 Set-Cookie（Session 已自动保存）；若响应体含 sess_key 则补存。"""
        if isinstance(data, dict):
            for key in ("sess_key", "SessionKey", "session_key", "sesskey"):
                value = data.get(key)
                if value:
                    self.session.cookies.set("sess_key", str(value))
                    return

    @staticmethod
    def _extract_list(data: Any) -> list[dict]:
        lst = None
        if isinstance(data, list):
            lst = data
        elif isinstance(data, dict):
            container = None
            for key in ("Data", "data", "result", "results", "list", "items", "rows"):
                if key in data:
                    container = data[key]
                    break
            if isinstance(container, list):
                lst = container
            elif isinstance(container, dict):
                for key in ("data", "result", "results", "list", "items", "rows"):
                    if isinstance(container.get(key), list):
                        lst = container[key]
                        break
        if not isinstance(lst, list):
            return []
        return [dict(e) for e in lst if isinstance(e, dict)]
